### ISSUE TITLE
[React 18] Possible Suspense Bug - Thrown error causes corrupted page

### ISSUE THREAD

**Issue by flybayer** [facebook/react#21802]
I've found what I think to be a React bug, although it could be something we are doing wrong inside Blitz.js

`React 18.0.0-alpha-ed6c091fe-20210701`

## Basic Explanation

The position of a top level, unneeded `<Suspense>` component next to an `<ErrorBoundary>` causes odd UI rendering or totally corrupted page.

Following are different variations of [this code](https://github.com/flybayer/blitz-react18/blob/main/app/pages/_app.tsx#L15-L22) inside `_app.tsx` (which is the same top level root as Next.js)

This code in `_app.tsx` causes a totally corrupted page (shows content on page that's no longer managed by React):
```tsx
    <Suspense fallback="Loading...">
      <ErrorBoundary FallbackComponent={RootErrorFallback}>
        <Component {...pageProps} />
      </ErrorBoundary>
    </Suspense>
```

This code in `_app.tsx` results in a correct end-result but has janky page transition.

```tsx
    <ErrorBoundary FallbackComponent={RootErrorFallback}>
      <Suspense fallback="Loading...">
        <Component {...pageProps} />
      </Suspense>
    </ErrorBoundary>
```

This code in `_app.tsx` without `<Suspense>` works flawless.

```tsx
    <ErrorBoundary FallbackComponent={RootErrorFallback}>
      <Component {...pageProps} />
    </ErrorBoundary>
```

I feel like this is a React bug because, to my knowledge, position of a `<Suspense>` component should have zero affect on a thrown error or ErrorBoundary.

## Further Explanation

In Blitz.js, we have a feature for restricting access to certain pages based on the user login status. We implement this [by throwing a custom `RedirectError`](https://github.com/blitz-js/blitz/blob/canary/packages/core/src/auth/auth-client.ts#L68-L74). And then inside our `<ErrorBoundary>` component, [we call `router.push()`](https://github.com/blitz-js/blitz/blob/canary/packages/core/src/error-boundary.tsx#L94-L98) when RedirectError is caught. And while the ErrorBoundary is in that temporary in-between state, we [render null](https://github.com/blitz-js/blitz/blob/canary/packages/core/src/error-boundary.tsx#L154-L156). The issue described above occurs when a page transition is initiated via RedirectError.

Please let me know if there's something wrong with this approach or if there's a better way to implement the pattern. 🙏 

## To Reproduce

1. Open this codesandbox: https://codesandbox.io/s/sharp-antonelli-u982e
1. Wait for the deps to install and home page to load (takes a bit)
1. See both the contents of `/protected` and `/` on the same page:
   - React dev tools shows that the `/protected` contents are no longer in the React tree
![image](https://user-images.githubusercontent.com/8813276/124835823-f3c33880-df4f-11eb-9371-e8d0aee570b9.png)

**Comment by gaearon** [comment]
Is there a chance you could reduce this to a minimal isolated sandbox with no dependencies? This usually helps a lot.

**Comment by flybayer** [comment]
@gaearon yes, I just updated the issue to have a minimal codesandbox. Sorry, I should have done that the first time around.

**Comment by gaearon** [comment]
To clarify, I'm asking if it's possible to make an isolated sandbox with no dependencies. E.g. just error boundary + Suspense + react. No frameworks or routers or other stuff. One file.

**Comment by gaearon** [comment]
Did this get fixed?

**Comment by flybayer** [comment]
We couldn't reproduce without nextjs

**Comment by gaearon** [comment]
Have you filed an issue against Next.js? We shouldn't lose track of this. It's ok to keep the issue open; somebody else may be able to isolate.

**Comment by flybayer** [comment]
@gaearon correction, we couldn't reproduce it without _blitz_. It worked fine with Next.js.

**Comment by gaearon** [comment]
Regardless, we should keep open if it looks like a bug in React

**Comment by Jikstra** [comment]
I think we had the same or a similiar bug in https://github.com/deltachat/deltachat-desktop 
We had an event that caused an error and every now and then a component would stay in the DOM, even if it wasn't part of the state anymore. We could fix it on our side (most likely, the bug is hard to reproduce) by fixing the unrelated thrown error.

**Comment by Jikstra** [comment]
> Hard to say anything without a reduced repro example.

We couldn't find any pattern on our side yet. It just happens sometimes. It just happend to me again, and suprisingly no error happend before. So it's probably another bug. 

Maybe it helps if i describe a bit our component hirarchy:
```
<MessageList>
  <MessagePage key=page0>
    <Message key=0>
    <Message key=1>
  <MessagePage key=page1>
    <Message key=2>
    <Message key=3>
</MessageList>
```

Both MessageList and MessagePage are memoized, (React.memo) and the only thing we can think of is that the memoization somehow forgets about them. MessagePages are getting rendered from a `state.messagePages.map`, adding a log statement there shows that the ghost MessagePages are indeed not part of the state anymore, but still present in the DOM. They also stay there as long as `<MessageList>` is not getting removed (we can switch to a gallery view which replaces that Component and then it cleans up that part of the DOM).

**Comment by gaearon** [comment]
Does this repro with stable React 18?
